Filter metadata when loading and more!

rstoneback opened the pull request:
# Description

Improves the consistency and accuracy of writing files across pandas and xarray. Adds in additional features for metadata handling. I've also included bug fixes in a couple areas that were relevant. There is a feedback update in _instrument.py, I needed to modify existing feedback to diagnose where fixes were needed, then I went ahead and fixed an issue relating to that code since I was there.

In some areas I changed some default values. I've lost sense on when/where/who in some areas. I went ahead and started at the strictest level, raising a DeprecationWarning, though not all of these may actually be warranted.

Addresses https://github.com/pysat/pysatNASA/issues/99, #938, #945, #946, #974, #986

 - Adds support for filtering metadata when loading. 
   - Includes a basic filter for the metadata added by pysat when creating netCDF4 files. 
   - Also includes a hook for developers to use to construct their own metadata filtering. 
 - Adds `meta_translation` kwarg to `to_netcdf` to support translating metadata labels when writing a file.
     - Adds `meta.to_translated_dict` function which makes this possible
     - Adds default export labels for fill_val to '_FillValue', and 'FillVal'
     - Corrects `inst.generic_meta_translator` which was using older meta standard
 - Adds support for dropping metadata labels during load
 - Adds support for translation table when loading netcdf files. This is done to enable `_FillValue' in the files in backwards
   compatible way.
 - Fixes #933, metadata check applied in more instances, and with improved feedback.
 - Prevents assignment of stale data paths in stored file information.

## Type of change

Please delete options that are not relevant.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide
instructions so we can reproduce. Please also list any relevant details for
your test configuration

- Testing by loading ICON IVM data and monitoring the results
- Wrote component tests for new functions
- Wrote integration tests
- Expanded number of test files written, loaded, modified, etc.
- Generated examples for documentation
- Developed jupyter notebook with a variety of file options, over in https://github.com/pysat/pysatTutorials/pull/7

**Test Configuration**:
* Operating system: MacOS
* Version number: Python 3.9.7
* Any details about your local setup that are relevant

# Checklist:

- [x] Make sure you are merging into the ``develop`` (not ``main``) branch
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] Add a note to ``CHANGELOG.md``, summarizing the changes

If this is a release PR, replace the first item of the above checklist with the release
checklist on the wiki: https://github.com/pysat/pysat/wiki/Checklist-for-Release

rstoneback commented:
Here are the requested flow charts for metadata and files.

<img width="1356" alt="image" src="https://user-images.githubusercontent.com/3595641/152395749-5fdb9acf-cdac-49a6-ab66-9345791f1027.png">

<img width="1249" alt="image" src="https://user-images.githubusercontent.com/3595641/152395828-5b9276cc-ada5-4351-959d-91bd6001497a.png">

rstoneback commented on pysat/_instrument.py:
```diff
@@ -523,6 +519,10 @@ def __init__(self, platform=None, name=None, tag='', inst_id='',
         # Store base attributes, used in particular by Meta class
         self._base_attr = dir(self)
 
+        # Run instrument init function, a basic pass function is used if the
```
Handeld more completelty in #987

rstoneback replied:
Also need the functionality in #987 to pass here. Going to go ahead and merge that in so testing can run.

rstoneback commented:
Don't forget the filter_netcdf function that removes all non-specified NaNs. This is left out of the diagram. Also left out of consideration when writing some testing. It was filtering out my NaNs!!!

rstoneback commented:
I have a plan to sort out the variety of features when writing a netCDF file. It involves deprecating a few existing functions. I will be reimplementing almost the same deprecated function, but usually with a different input or so. Will be moving towards a standard dict, rather than operating on Meta itself. Meta is good for users at the user level. Not so good when at the about to write a file level.

rstoneback commented:
@jklenzing I may need some help from you getting `test_read_netcdf4_w_epoch_kwargs` going post merge. Setting epoch_name when writing a file takes the main 'time' index in xarray and stores it in the file as epoch_name, same as with the pandas format. When loading back up, those dates are already in datetime format. The `xr.DataArray(pds.to_datetime(data[epoch_name], ...` line is apparently unhappy with datetimes as input. I get here by commenting lines 283-292 in test_utills_io.py. I also tried putting in `data[epoch_name].values.astype(int)` but it reported it didn't like 'ms' for units.

rstoneback commented:
> @jklenzing I may need some help from you getting `test_read_netcdf4_w_epoch_kwargs` going post merge. Setting epoch_name when writing a file takes the main 'time' index in xarray and stores it in the file as epoch_name, same as with the pandas format. When loading back up, those dates are already in datetime format. The `xr.DataArray(pds.to_datetime(data[epoch_name], ...` line is apparently unhappy with datetimes as input. I get here by commenting lines 283-292 in test_utills_io.py. I also tried putting in `data[epoch_name].values.astype(int)` but it reported it didn't like 'ms' for units.

Could be I need to update the values when time is written. xarray is currently handling it.... Sounds like an area for me to check with fresh eyes tomorrow.

rstoneback commented:
Documentation is not quite done, but the code works and is in as good a shape as time allows.

jklenzing commented on pysat/utils/io.py:
```diff
@@ -724,47 +1294,71 @@ def load_netcdf_xarray(fnames, strict_meta=False, file_format='NETCDF4',
-                                                        origin=epoch_origin),
-                                        coords=data[epoch_name].coords)
+    # Rename `epoch_name` to 'time'.
+    if epoch_name != 'time':
+        if epoch_name in data.dims:
+            data = data.rename({epoch_name: 'time'})
+        else:
+            estr = ''.join(['User provided time information variable ',
+                            epoch_name, ' not found in loaded data ',
+                            'dimensions ', repr(data.dims)])
+            raise ValueError(estr)
+
```
Time information is not always stored in `data.dims`.  For example, in GOLD NMAX data, the dims are basically integer indices, with time and geographic info stored as variables.

rstoneback replied:
Thanks for the note. I've generalized the check.   

🐬 Don't let GOLD push you around. While I haven't looked, data can usually be tweaked 🐵

rstoneback replied:
I expanded the metadata assignment from `data.variables` to a combo of `variables`, `coords`, and `dims`. Hopefully that helps out. 👍

rstoneback commented on pysat/utils/io.py:
```diff
@@ -396,7 +862,7 @@ def load_netcdf_pandas(fnames, strict_meta=False, file_format='NETCDF4',
         'NETCDF3_CLASSIC', 'NETCDF3_64BIT', 'NETCDF4_CLASSIC', or 'NETCDF4'.
         (default='NETCDF4')
-    epoch_name : str
+    epoch_name : str or NoneType
```
Should the default be updated to None to trigger the warning below. Warning is triggered by default when calling from `load_netcdf`

aburrell replied:
No strong feelings.

rstoneback replied:
I've confirmed that `load_netcdf` default is still None. Also, `load_netcdf_pandas` didn't exist in previous versions. Which means no history of defaults to worry about at this level.

rstoneback commented on pysat/utils/io.py:
```diff
@@ -382,7 +846,9 @@ def load_netcdf_pandas(fnames, strict_meta=False, file_format='NETCDF4',
                                'min_val': ('value_min', np.float64),
                                'max_val': ('value_max', np.float64),
-                               'fill_val': ('fill', np.float64)}):
+                               'fill_val': ('fill', np.float64)},
```
The labels class default may need to be updated. Current default has more labels than standard pysat default. So writing and loading a file technically changes metadata.

aburrell replied:
Sounds like this should be an issue.

rstoneback replied:
I unclearly was talking the definition right here. I can post an issue, or just, whoop whoop, take care of it here.

aburrell replied:
Let's not make this PR any bigger 🐳

rstoneback replied:
Just big bones. 

That comment is a bit stale. Since I posted that I made an issue.

rstoneback commented:
Thanks for the comments so far @jklenzing. I have rounded out the documentation, updated pull request info, etc. Should be  fully reviewable now. Apologies for pushing the review button a bit early compared to those things.

aburrell commented:
Postponing review until conflicts resolved.

rstoneback commented:
Thanks. Didn't see that there. My webpage had gone stale.

aburrell commented on pysat/tests/test_utils_io.py:
```diff
@@ -374,6 +385,100 @@ def test_netcdf_attribute_override(self):
+        # Write file.
+        io.inst_to_netcdf(self.testInst, fname=outfile, export_nan=export_nan)
+
+        if drop_labels:
+            drop_list = [drop_label]
+        else:
+            drop_list = []
```
```suggestion
        drop_list = [drop_label] if drop_labels else []
```

aburrell commented on pysat/tests/test_utils_io.py:
```diff
@@ -680,75 +796,350 @@ def test_add_netcdf4_standards_to_meta(self, caplog, missing):
+            for var in open_f.variables.keys():
+                test_vars = open_f[var].ncattrs()
+                # Confirm translated labels are in the file
+                # Avoid time variables
```
```suggestion
                # and avoid time variables
```

aburrell commented on pysat/tests/test_utils_io.py:
```diff
@@ -680,75 +796,350 @@ def test_add_netcdf4_standards_to_meta(self, caplog, missing):
+                    form = open_f[var].getncattr('Format')
+                    for key in meta_trans.keys():
+                        # String data doesn't have fill
+                        if key != 'fill' and (form != 'S1'):
```
```suggestion
                        if key != 'fill' and form != 'S1':
```

rstoneback replied:
My experience has been that the second set needs parentheses.

aburrell commented on pysat/tests/test_utils_io.py:
```diff
@@ -680,75 +796,350 @@ def test_add_netcdf4_standards_to_meta(self, caplog, missing):
+        return
+
+    @pytest.mark.parametrize('assign_flag', [True, False])
+    def test_meta_processor_to_from_netcdf4(self, assign_flag):
```
Confused about the nesting of functions in this test method.  Why are they not at the class level?

rstoneback replied:
I'm testing the meta_processor functionality, and that requires providing a function to the meta_processor keyword. The internal functions `to_meta_proc` and `from_meta_proc` only have relevance within the larger test, `test_meta_processor....`. The meta_processor functionality itself doesn't do much, it is just supposed to provide metadata input to the supplied function as well as accept the returned dict. So I test within the `to_...` and `from` to ensure things are a dict, make changes, and then outside that function ensure the changes are applied.

aburrell replied:
I thought that might be the case.  Personally, I'd prefer to have these defined at the class level, as it makes the code easier to read and if, for some reason, another test ends up needing these functions they're easier to find.  It's a style preference, though, so it's your call.

rstoneback replied:
Updated.

aburrell commented on pysat/tests/test_utils_io.py:
```diff
@@ -829,3 +1207,405 @@ def test_pysat_meta_to_xarray_attr(self, export_nan):
+        for var in self.out.keys():
+            for tvar in np.arange(4):
+                tstr = 'array_test{:1d}'.format(tvar)
+                assert tstr in self.out[var], estr
+                assert self.out[var][tstr] == tvar + 1, estr2
```
Error strings should provide more information about the variables being compared.

aburrell commented on pysat/utils/io.py:
```diff
@@ -172,121 +167,555 @@ def filter_netcdf4_metadata(inst, mdata_dict, coltype, remove=False,
+
+                # Filter metadata
+                remove = True if sctype == str else False
+                out_meta_dict[sname] = \
+                    filter_netcdf4_metadata(inst, out_meta_dict[sname],
```
Recommend breaking on open paren

aburrell commented on pysat/utils/io.py:
```diff
@@ -172,121 +167,555 @@ def filter_netcdf4_metadata(inst, mdata_dict, coltype, remove=False,
+
+            # Filter metdata for other netCDF4 requirements
+            remove = True if index_type == str else False
+            out_meta_dict[lower_var] = \
+                filter_netcdf4_metadata(inst, out_meta_dict[lower_var],
```
Break on open paren

aburrell commented on pysat/utils/io.py:
```diff
@@ -172,121 +167,555 @@ def filter_netcdf4_metadata(inst, mdata_dict, coltype, remove=False,
+
+            # Filter metdata for other netCDF4 requirements
+            remove = True if coltype == str else False
+            out_meta_dict[lower_var] = \
+                filter_netcdf4_metadata(inst, out_meta_dict[lower_var],
```
```suggestion
            out_meta_dict[lower_var] = filter_netcdf4_metadata(
                inst, out_meta_dict[lower_var],
```

aburrell commented on pysat/utils/io.py:
```diff
@@ -172,121 +167,555 @@ def filter_netcdf4_metadata(inst, mdata_dict, coltype, remove=False,
+
+        if 'meta' in lower_sub_keys:
+            # Higher dimensional data, recursive treatment.
+            mdict[key]['meta'] = remove_netcdf4_standards_from_meta(mdict[key]
+                                                                    ['meta'],
+                                                                    '',
+                                                                    labels)
```
```suggestion
            mdict[key]['meta'] = remove_netcdf4_standards_from_meta(
                mdict[key]['meta'], '', labels)
```

aburrell commented on pysat/utils/io.py:
```diff
@@ -172,121 +167,555 @@ def filter_netcdf4_metadata(inst, mdata_dict, coltype, remove=False,
-                    smeta_dict = {'Depend_0': epoch_name,
-                                  'Depend_1': obj_dim_names[-1],
-                                  'Display_Type': 'Spectrogram',
-                                  'Format': inst._get_var_type_code(scoltype),
-                                  'Var_Type': 'data'}
-                    inst.meta[svar] = smeta_dict
+    # Update labels required by netCDF4
+    trans_table['_FillValue'] = meta.labels.fill_val
+    trans_table['FillVal'] = meta.labels.fill_val
```
Recommend combining these lines into one init statement.

aburrell commented on pysat/utils/io.py:
```diff
@@ -172,121 +167,555 @@ def filter_netcdf4_metadata(inst, mdata_dict, coltype, remove=False,
+    trans_table['_FillValue'] = meta.labels.fill_val
+    trans_table['FillVal'] = meta.labels.fill_val
+
+    # We *may* need to keep this for backwards compatibility. Unintended use
+    # of 'fill' in pysat generated files.
+    trans_table['fill'] = meta.labels.fill_val
```
If this is not sure, create an issue and add TODO for easier tracking.

rstoneback replied:
I think it came with the addition/expansion of `pysat.utils.io`. If that was this version then we don't need to keep the 'fill' part.

aburrell replied:
>  If that was this version then we don't need to keep the 'fill' part.

I don't understand this sentence very well.  Are you saying this line can now be removed?

rstoneback replied:
I double checked `main`. Wrote a test file and 'fill' is in the default metadata. I will update the comment.

aburrell commented on pysat/utils/io.py:
```diff
@@ -172,121 +167,555 @@ def filter_netcdf4_metadata(inst, mdata_dict, coltype, remove=False,
+
+    """
+
+    # Define a default translation
```
Init dict with line 521 or something like `trans_table = {val: [val] for val in inst.meta.labels.label_attrs.keys()}`

aburrell commented on pysat/utils/io.py:
```diff
@@ -172,121 +167,555 @@ def filter_netcdf4_metadata(inst, mdata_dict, coltype, remove=False,
+        trans_table = default_to_netcdf_translation_table(inst)
+
+    # Confirm there are no duplicated translation labels
+    trans_labels = []
+    for key in trans_table.keys():
+        trans_labels.extend(trans_table[key])
```
Do this cleaner in one line.

aburrell commented on pysat/utils/io.py:
```diff
@@ -172,121 +167,555 @@ def filter_netcdf4_metadata(inst, mdata_dict, coltype, remove=False,
+        # Iterate over metadata for given variable `var_key`
+        for file_key in meta_dict[var_key].keys():
+            # Apply translation if defined
+            if file_key in trans_table:
+                new_key = trans_table[file_key]
+            else:
+                new_key = file_key
```
Possible to do in one line, not sure if it would be too long, though.

rstoneback replied:
Just a bit too long.

aburrell commented on pysat/utils/io.py:
```diff
@@ -172,121 +167,555 @@ def filter_netcdf4_metadata(inst, mdata_dict, coltype, remove=False,
+                        check = check1 and check2
+                    except TypeError:
+                        check = True
+                    if check:
```
```suggestion

                    if check:
```

aburrell commented on pysat/utils/io.py:
```diff
@@ -172,121 +167,555 @@ def filter_netcdf4_metadata(inst, mdata_dict, coltype, remove=False,
+def meta_array_expander(meta_dict):
+    """Expand meta arrays by storing each element with new incremented label.
+
+    if meta_dict[variable]['label'] = [ item1, item2, ..., itemn] then
+    the returned dict will contain, meta_dict[variable]['label0'] = item1,
+    meta_dict[variable]['label1'] = item2,
+    meta_dict[variable]['labeln-1'] = itemn
+
```
```suggestion
    If `meta_dict[variable]['label'] = [ item1, item2, ..., itemn]` then
    the returned dict will contain: `meta_dict[variable]['label0'] = item1`,
    `meta_dict[variable]['label1'] = item2`, and so on up to
    `meta_dict[variable]['labeln-1'] = itemn`.

```

aburrell commented on pysat/utils/io.py:
```diff
@@ -782,10 +1414,90 @@ def load_netcdf_xarray(fnames, strict_meta=False, file_format='NETCDF4',
+
+    Parameters
+    ----------
+    data : xarray.Dataset
```
Missing description

aburrell commented on pysat/utils/io.py:
```diff
@@ -782,10 +1414,90 @@ def load_netcdf_xarray(fnames, strict_meta=False, file_format='NETCDF4',
+
+    Returns
+    -------
+    variables :
```
Missing type(s).  Also doesn't match return in all cases.

Docstring is also missing Raises section.

jklenzing commented:
This branch is breaking a  number of things in the pysatNASA develop branch. Recurring errors through instruments:
```
KeyError: 'Epoch'
ValueError: User provided time information variable Epoch not found in lo...
ValueError: ('Lengths must match to compare', (86400,), (1,))
```

jklenzing commented:
Tracking down which errors are unique to this branch. Documenting existing errors in pysatNASA issues.

KeyError: 'Epoch' is being triggered in this branch by the ICON xarray instruments.
```
~/code/core/pysat/pysat/instruments/methods/general.py in convert_timestamp_to_datetime(inst, sec_mult, epoch_name)
    176                   DeprecationWarning, stacklevel=2)
    177 
--> 178     if inst.data[epoch_name].dtype == 'float64':
    179         inst.data[epoch_name] = pds.to_datetime(
    180             [dt.datetime.utcfromtimestamp(int(np.floor(epoch_time * sec_mult)))
```

jklenzing commented:
The ICON issue is fixed by updating the default epoch name in a deprecated function.  I've pushed the fix.  The GOLD dataset still needs to be updated, which probably requires some development discussion, since it doesn't fit the assumptions that pysat makes.

rstoneback commented:
> The ICON issue is fixed by updating the default epoch name in a deprecated function. I've pushed the fix. The GOLD dataset still needs to be updated, which probably requires some development discussion, since it doesn't fit the assumptions that pysat makes.

Yeah, this branch looks for 'time' for xarray Instrument instead of 'Epoch'. Thanks for pushing a fix.

rstoneback commented on pysat/tests/classes/cls_instrument_integration.py:
```diff
@@ -0,0 +1,81 @@
+"""Tests for pysat.Instrument integration tests.
+
+Includes:
+* Ensuring no stale data paths after changing pysat.params['data_dirs']
```
I copied a format from a different test. This integration test class could collect a variety of tests and an index should make it easier to identify what is in the file.

rstoneback commented on pysat/_meta.py:
```diff
@@ -1467,6 +1477,57 @@ def add_epoch_metadata(self, epoch_name):
+
+        # First Order Data
+        for key in self.data.index:
+            if preserve_case:
+                case_key = self.var_case_name(key)
+            else:
+                case_key = key.lower()
```
The only place there is something akin to this if check is in the `to_dict` method. 2 of the 3 uses in there will be removed with the removal of higher order metadata. I added a TODO note.

rstoneback commented:
I added a function, `xarray_all_vars`, to make the `load_netcdf` more broadly compatible. I also iterate over all_vars for metadata, as well as when searching for a supplied `epoch_name`. I replaced the originating code for `xarray_all_vars` with the function call. Turns out the original code had a bug, and double added dims instead of coordinates.  While building tests, realized we needed some robustness if `time` already existed in the file. Anyway, the recent commits cover `xarray_all_vars`, tests, and simple formatting of comments, etc.

aburrell commented on docs/tutorial/tutorial_files.rst:
```diff
@@ -238,22 +359,151 @@ information on adding a new dataset to pysat, see :ref:`rst_new_inst`.
+    # while the values are the labels from the file and type.
+    # The labels are applied last in the loading process.
+    # Only one type is currently supported for each metadata label.
+    labels = {'units': ('Units', str), 'name': ('Long_Name', str),
+             'notes': ('Var_Notes', str), 'desc': ('CatDesc', str),
+             'plot': ('plot', str), 'axis': ('axis', str),
+             'scale': ('scale', str),
+             'min_val': ('Valid_Min', np.float64),
+             'max_val': ('Valid_Max', np.float64),
+             'fill_val': ('fill', np.float64)}
```
Is this alignment correct?  It looks off by one space here.

aburrell commented on docs/tutorial/tutorial_files.rst:
```diff
@@ -238,22 +359,151 @@ information on adding a new dataset to pysat, see :ref:`rst_new_inst`.
+    # while the values are the labels from the file and type.
+    # The labels are applied last in the loading process.
+    # Only one type is currently supported for each metadata label.
+    local_labels = {'units': ('UNITS', str), 'name': ('LongEST_Name', str),
+             'notes': ('FLY', str), 'desc': ('DIGits', str),
+             'plot': ('plottER', str), 'axis': ('axisER', str),
+             'scale': ('scalER', str),
+             'min_val': ('INVALIDmin', np.float64),
+             'max_val': ('invalidMAX', np.float64),
+             'fill_val': ('fillerest', np.float64)}
```
Fix alignment.

aburrell commented on pysat/instruments/methods/testing.py:
```diff
@@ -102,72 +103,113 @@ def initialize_test_meta(epoch_name, data_keys):
+                        'value_max': 4294967295, 'fill': -1,
+                        'desc': ''.join(['Testing profile multi-dimensional ',
+                                         'data indexed by time.']),
+                        'notes': ''.join(['Note the value_max is largest ',
+                                          'netCDF4 supports, but is ',
+                                          'lower than actual 64-bit int limit.',
+                                          ])}
```
Stylistically, I don't like this extra comma and hanging closing parens.  Recommend breaking the text up a bit more or breaking the lasts join list on the opening square bracket.

aburrell commented on pysat/instruments/methods/testing.py:
```diff
@@ -102,72 +103,113 @@ def initialize_test_meta(epoch_name, data_keys):
+                                    'desc': ''.join(['Simulated fractional O+ ',
+                                                     'composition.'])}
+    meta['alt_profiles'] = {'value_min': 0., 'value_max': 25., 'fill': np.nan,
+                            'desc': ''.join(['Testing profile multi-dimensiona',
+                                             'l data indexed by float.']),
```
Recommend not breaking up words for ease of reading.

aburrell commented on pysat/instruments/pysat_testing.py:
```diff
@@ -61,24 +60,23 @@ def load(fnames, tag='', inst_id='', sim_multi_file_right=False,
     num_samples : int
         Maximum number of times to generate.  Data points will not go beyond the
         current day. (default=86400)
     test_load_kwarg : any or NoneType
```
```suggestion
    test_load_kwarg : any
```

aburrell commented on pysat/instruments/pysat_testing.py:
```diff
@@ -61,24 +60,23 @@ def load(fnames, tag='', inst_id='', sim_multi_file_right=False,
         current day. (default=86400)
     test_load_kwarg : any or NoneType
-        Testing keyword (default=None)
+        Testing keyword. (default=None)
```
Not descriptive, please expand.

aburrell commented on pysat/instruments/pysat_testing2d.py:
```diff
@@ -97,108 +97,135 @@ def load(fnames, tag='', inst_id='', malformed_index=False,
-    # use that info to create a signal that is continuous from that start
-    # going to presume there are 5820 seconds per orbit (97 minute period)
+    # Need to create simple orbits here. Have start of first orbit
+    # at 2009,1, 0 UT. 14.84 orbits per day.
+    # Figure out how far in time from the root start a measurement is and
```
```suggestion
    # at 2009,1, 0 UT. 14.84 orbits per day. Figure out how far in time from the root start a measurement is and
```

aburrell commented on pysat/instruments/pysat_testing2d.py:
```diff
@@ -97,108 +97,135 @@ def load(fnames, tag='', inst_id='', malformed_index=False,
                                              period=iperiod['lt'],
                                              data_range=drange['lt'])
-    # do slt, 20 second offset from mlt
+    # SLT, 20 second offset from `mlt`.
```
```suggestion

    # SLT, 20 second offset from `mlt`.
```

aburrell commented on pysat/instruments/pysat_testing2d.py:
```diff
@@ -97,108 +97,135 @@ def load(fnames, tag='', inst_id='', malformed_index=False,
 
-    # create a few simulated profiles
-    # DataFrame at each time with mixed variables
+    # Create a few simulated profiles.
+    # DataFrame at each time with mixed variables.
```
```suggestion
    # Create a few simulated profiles. This results in a pds.DataFrame at each time with mixed variables.
```

aburrell commented on pysat/instruments/pysat_testing2d_xarray.py:
```diff
@@ -25,15 +25,15 @@
-# Init method
+# Init method.
 init = mm_test.init
 
 
```
Too many spaces?

aburrell commented on pysat/instruments/pysat_testing2d_xarray.py:
```diff
@@ -57,13 +57,12 @@ def load(fnames, tag='', inst_id='', malformed_index=False,
         Maximum number of times to generate.  Data points will not go beyond the
         current day. (default=864)
     test_load_kwarg : any or NoneType
-        Testing keyword (default=None)
+        Testing keyword. (default=None)
```
Same as previous.

aburrell commented on pysat/instruments/pysat_testing2d_xarray.py:
```diff
@@ -90,82 +89,100 @@ def load(fnames, tag='', inst_id='', malformed_index=False,
-    # use that info to create a signal that is continuous from that start
-    # going to presume there are 5820 seconds per orbit (97 minute period)
+    # Need to create simple orbits here. Have start of first orbit
+    # at 2009,1, 0 UT. 14.84 orbits per day.
+    # Figure out how far in time from the root start a measurement is and
```
```suggestion
    # at 2009,1, 0 UT. 14.84 orbits per day. Figure out how far in time from the root start a measurement is and
```

aburrell commented on pysat/instruments/pysat_testing2d_xarray.py:
```diff
@@ -90,82 +89,100 @@ def load(fnames, tag='', inst_id='', malformed_index=False,
-    # sat moves at 360/5820 deg/s, Earth rotates at 360/86400, takes extra time
-    # to go around full longitude
+    # Create a fake satellite longitude, resets every 6240 seconds.
+    # Sat moves at 360/5820 deg/s, Earth rotates at 360/86400, takes extra time
```
```suggestion
    # Satellite moves at 360/5820 deg/s, Earth rotates at 360/86400 deg/s, takes extra time
```

aburrell commented on pysat/instruments/pysat_testing2d_xarray.py:
```diff
@@ -90,82 +89,100 @@ def load(fnames, tag='', inst_id='', malformed_index=False,
+                           np.array([1] * len(data.indexes[epoch_name]),
+                           dtype=np.int16))
+    data['int32_dummy'] = ((epoch_name),
+                           np.array([1] * len(data.indexes[epoch_name]),
```
```suggestion
                           np.array([1] * len(data.indexes[epoch_name]),
```

Would be clearer to use `np.ones`

aburrell commented on pysat/utils/io.py:
```diff
@@ -782,10 +1405,128 @@ def load_netcdf_xarray(fnames, strict_meta=False, file_format='NETCDF4',
+
+    """
+
+    # Construct a list of all variables, dimensions, and coordinates.
+
+    # Start with the dimensions!
```
```suggestion
    # Construct a list of all variables, dimensions, and coordinates.
    # Start by obtaining the dimensions.
```

rstoneback replied:
![](https://i.imgflip.com/6fszqz.jpg)

Plus what would Nathaniel Hawthorne say?!

aburrell commented on pysat/utils/io.py:
```diff
@@ -836,13 +1578,26 @@ def inst_to_netcdf(inst, fname, base_instrument=None, epoch_name='Epoch',
+    meta_translation : dict or NoneType
+        The keys in the input dict are used to map
+        metadata labels for `inst` to one or more values used when writing
+        the file. eg. {meta.labels.fill_val: ['FillVal', '_FillValue']} would
```
```suggestion
        the file. E.g., `{meta.labels.fill_val: ['FillVal', '_FillValue']}` would
```

aburrell commented on pysat/instruments/pysat_testing2d.py:
```diff
@@ -97,108 +98,136 @@ def load(fnames, tag='', inst_id='', malformed_index=False,
-    # this time signal used for 2D profiles associated with each time in main
-    # DataFrame
+
+    # Higher rate time signal (for scalar >= 2).
+    # This time signal is used for 2D profiles associated with each time in main
+    # DataFrame.
```
```suggestion
    # Higher rate time signal (for scalar >= 2). This time signal is used
    # for 2D profiles associated with each time in main DataFrame.
```

aburrell commented on pysat/instruments/pysat_testing_xarray.py:
```diff
@@ -127,57 +129,61 @@ def load(fnames, tag='', inst_id='', sim_multi_file_right=False,
-    # create a fake longitude, resets every 6240 seconds
-    # sat moves at 360/5820 deg/s, Earth rotates at 360/86400, takes extra time
-    # to go around full longitude
+    # Create a fake longitude, resets every 6240 seconds.
+    # Sat moves at 360/5820 deg/s, Earth rotates at 360/86400, takes extra time
+    # to go around full longitude.
```
```suggestion
    # Create a fake longitude, resets every 6240 seconds. Sat moves at
    # 360/5820 deg/s, Earth rotates at 360/86400, takes extra time to go
    # around full longitude.
```

aburrell commented on pysat/instruments/pysat_testmodel.py:
```diff
@@ -67,25 +65,26 @@ def load(fnames, tag='', inst_id='', start_time=None, num_samples=96,
 
-    # Create an artificial model data set
+    # Create an artificial model data set.
     if tag == '':
         freq_str = '900S'
     else:
         freq_str = '1H'
```
```suggestion
    freq_str = '900S' if tag == '' else '1H'
```

aburrell commented on pysat/utils/io.py:
```diff
@@ -17,69 +16,62 @@
+    if epoch_name in pysat_meta.keys():
+        for meta_key in pysat_meta[epoch_name].keys():
+            # Assign attributes
+            xr_data[epoch_name].attrs[meta_key] = \
+                pysat_meta[epoch_name][meta_key]
```
Break on open `[`

aburrell commented on pysat/utils/io.py:
```diff
@@ -172,121 +167,532 @@ def filter_netcdf4_metadata(inst, mdata_dict, coltype, remove=False,
+
+
+def default_to_netcdf_translation_table(inst):
+    """Return metadata translation table with minimal netcdf requirements.
```
```suggestion
    """Create a metadata translation table with minimal netCDF requirements.
```

For the docstrings in this file, recommend using "Create" or "Extract", etc instead of "Return", since the latter is technically covered by the Returns section and the former give more insight into the innards of the function.

rstoneback commented:
Thanks @aburrell!! Suggestions incorporated. Ready to roll! 🐬

aburrell commented on pysat/instruments/pysat_testing2d.py:
```diff
@@ -110,95 +111,122 @@ def load(fnames, tag='', inst_id='', malformed_index=False,
-    # sat moves at 360/5820 deg/s, Earth rotates at 360/86400, takes extra time
-    # to go around full longitude
+
+    # Create a fake longitude, resets every 6240 seconds.
+    # Aat moves at 360/5820 deg/s, Earth rotates at 360/86400, takes extra time
+    # to go around full longitude.
```
```suggestion
    # Create a fake longitude, resets every 6240 seconds. Sat moves at
    # 360/5820 deg/s, Earth rotates at 360/86400, takes extra time to go
    # around full longitude.
```

aburrell commented on pysat/instruments/pysat_testmodel.py:
```diff
@@ -94,18 +90,18 @@ def load(fnames, tag='', inst_id='', start_time=None, num_samples=96,
 
-        # Simulate altitude values at model points
-        # Initiliaze memory
+        # Simulate altitude values at model points.
+        # Initialize memory.
```
```suggestion
        # Simulate altitude values at the model nodes.
        # Start by initializing the memory.
```

aburrell commented on pysat/utils/io.py:
```diff
@@ -950,7 +1721,7 @@ def inst_to_netcdf(inst, fname, base_instrument=None, epoch_name='Epoch',
             attrb_dict[akey] = int(attrb_dict[akey])
 
     # Check if there are multiple variables with same characters
```
```suggestion
    # Check if there are multiple variables with same characters,
```

aburrell commented on pysat/utils/io.py:
```diff
@@ -950,7 +1721,7 @@ def inst_to_netcdf(inst, fname, base_instrument=None, epoch_name='Epoch',
 
     # Check if there are multiple variables with same characters
-    # but with different case
+    # but with different case.
```
```suggestion
    # but with a different case.
```

aburrell reviewed (commented):
Don't forget to re-request a review from me when needed.  I don't always read the comment emails.

jklenzing reviewed (approved):
Approving my chunk of comments.  @aburrell may still have updates

Verified this works for the pysatModels RC and pysat/pysatNASA#110

aburrell commented on pysat/utils/io.py:
```diff
@@ -172,121 +167,532 @@ def filter_netcdf4_metadata(inst, mdata_dict, coltype, remove=False,
+
+    Note
+    ----
+    pysat.Meta can not take array data.
```
```suggestion
    pysat.Meta can not take array-like or list-like data.
```

aburrell commented on pysat/utils/io.py:
```diff
@@ -500,9 +990,9 @@ def load_netcdf_pandas(fnames, strict_meta=False, file_format='NETCDF4',
                     obj_key = dim[0]
                 else:
                     estr = ''.join(['Epoch label: "', epoch_name, '"',
-                                    ' was not found in loaded dimensions [',
+                                    ' not found in loaded dimensions [',
                                     ', '.join(dim), ']'])
-                    raise KeyError(estr)
+                    raise ValueError(estr)
```
I don't like the new wording, keeping "was" seems to create an easier to read sentence.  Also curios as to the motivation behind the Key->Value change.

rstoneback replied:
For consistency with another equivalent error, in xarray I think. Based on the pysatModels SAMI test looks like I went the wrong way. 🙈 I'll spin up on this and the last remaining comment way up there about a None here shortly (today).

rstoneback replied:
To close the loop, Errors were updated for consistency with the existing statement. The SAMI test error was something else. Fix incorporated into pysatModels develop.

rstoneback commented on pysat/utils/io.py:
```diff
@@ -724,68 +1278,254 @@ def load_netcdf_xarray(fnames, strict_meta=False, file_format='NETCDF4',
+                                ' is not a dimension.'])
+                pysat.logger.warning(wstr)
+            else:
+                estr = ''.join(['Epoch label: "', epoch_name, '"',
```
This was the one. Close enough to pass the same test. Will improve. Since the pandas ones already existed I'll use them as the reference.

aburrell commented on pysat/tests/test_utils_io.py:
```diff
@@ -151,37 +159,54 @@ def test_basic_write_and_read_netcdf_mixed_case_meta_format(self):
+            update_files=True)
+
+        # Confirm data path is correct.
+        assert os.path.normpath(netcdf_inst.files.data_path)\
```
```suggestion
        assert os.path.normpath(netcdf_inst.files.data_path) \
```

aburrell commented on pysat/tests/test_utils_io.py:
```diff
@@ -680,75 +865,358 @@ def test_add_netcdf4_standards_to_meta(self, caplog, missing):
+        return
+
+    def meta_proc_stub(self, meta_dict, vals=None, remove_labels=None):
+        """`meta_processor` function for load/write processor tests.
```
```suggestion
        """Load and write processor tests; supports `meta_processor`.
```

aburrell commented on pysat/tests/test_utils_io.py:
```diff
@@ -829,3 +1284,436 @@ def test_pysat_meta_to_xarray_attr(self, export_nan):
+    def test_xarray_vars_no_time(self, time_label):
+        """Test `xarray_vars_no_time`.
+
+        time_label : str
```
```suggestion
        Parameters
        ------------
        time_label : str
```

aburrell commented on pysat/utils/io.py:
```diff
@@ -172,121 +167,532 @@ def filter_netcdf4_metadata(inst, mdata_dict, coltype, remove=False,
+
+    Note
+    ----
+    Does perform filtering to remove variable metadata not supported by the
+    SPDF ISTP/IACG NetCDF standards using
```
```suggestion
    Removes variable metadata not supported by the SPDF ISTP/IACG
    NetCDF standards using
```

aburrell commented on pysat/utils/io.py:
```diff
@@ -172,121 +167,532 @@ def filter_netcdf4_metadata(inst, mdata_dict, coltype, remove=False,
+    ----
+    Does perform filtering to remove variable metadata not supported by the
+    SPDF ISTP/IACG NetCDF standards using
+    `pysat.utils.io.filter_netcdf4_metadata`.
```
Recommend a 'See Also' section that points to this function.

aburrell commented on pysat/utils/io.py:
```diff
@@ -172,121 +167,532 @@ def filter_netcdf4_metadata(inst, mdata_dict, coltype, remove=False,
+                # for either a Series or a DataFrame.
+                dims = (dims[0], 0)
+
+            for i, dim in enumerate(dims[:-1]):
+                # Don't need to go over last dimension value,
+                # it covers number of columns (if a DataFrame)
+                obj_dim_names.append(var)
```
Why is enumerate being used?  It is slower and the variable `i` is not called.

aburrell commented on pysat/utils/io.py:
```diff
@@ -172,121 +167,532 @@ def filter_netcdf4_metadata(inst, mdata_dict, coltype, remove=False,
+
+    Returns
+    -------
+    dict
```
```suggestion
    trans_table : dict
```

aburrell commented on pysat/utils/io.py:
```diff
@@ -724,68 +1278,254 @@ def load_netcdf_xarray(fnames, strict_meta=False, file_format='NETCDF4',
+        estr = ''.join(["Didn't find time dimension '", time_label, "'"])
+        raise ValueError(estr)
+
+    return []
```
This doesn't jive with the return statement.    Also, it looks like it is never reached.  Recommend replacing earlier `return` with `break` for clarity.

aburrell commented on pysat/utils/io.py:
```diff
@@ -724,68 +1278,254 @@ def load_netcdf_xarray(fnames, strict_meta=False, file_format='NETCDF4',
+
+    Returns
+    -------
+    variables : list
```
```suggestion
    vars : list
```

aburrell commented on pysat/utils/io.py:
```diff
@@ -724,68 +1278,254 @@ def load_netcdf_xarray(fnames, strict_meta=False, file_format='NETCDF4',
+def xarray_all_vars(data):
+    """Extract all variable names, including dimensions and coordinates.
+
+    data : xarray.Dataset
```
Missing Parameters again.  Recommend checking over all docstrings.

jklenzing commented:
Meanwhile, at pysatNASA....
![Screen Shot 2022-05-23 at 10 00 54 AM](https://user-images.githubusercontent.com/19592220/169836881-c45c34b9-f878-41e7-8f9f-5f56b77fb002.png)

aburrell commented on pysat/tests/test_utils_io.py:
```diff
@@ -90,17 +97,18 @@ def eval_loaded_data(self, test_case=True):
+            # Don't apply to 'time'
+            xarr_vars = io.xarray_vars_no_time(self.testInst.data)
             map_keys = {dkey: dkey.upper()
-                        for dkey in self.testInst.data.data_vars.keys()}
+                        for dkey in xarr_vars}
```
This can be on one line now, most likely.

aburrell commented on pysat/tests/test_utils_io.py:
```diff
@@ -139,9 +146,10 @@ def test_basic_write_and_read_netcdf_mixed_case_meta_format(self):
+            xarr_vars = io.xarray_vars_no_time(self.testInst.data)
             self.testInst.data = self.testInst.data.rename(
                 {dkey: dkey.upper()
-                 for dkey in self.testInst.data.data_vars.keys()})
+                 for dkey in xarr_vars})
```
Also here.

aburrell commented on pysat/tests/test_utils_io.py:
```diff
@@ -151,37 +159,54 @@ def test_basic_write_and_read_netcdf_mixed_case_meta_format(self):
+        # Load the written file directly into an Instrument.
         netcdf_inst = pysat.Instrument(
             'pysat', 'netcdf', data_dir=file_path,
-            file_format=file_root, pandas_format=self.testInst.pandas_format)
+            file_format=file_root, pandas_format=self.testInst.pandas_format,
+            update_files=True)
```
You can move `update_files` to line 185 and reduce the number of lines.

aburrell commented on pysat/tests/test_utils_io.py:
```diff
@@ -214,29 +239,116 @@ def test_inst_write_and_read_netcdf(self):
-        del netcdf_inst
+        del netcdf_inst.data, netcdf_inst
+
+        # Delete the file. This doesn't work on Windows. It does work if
+        # executed before the netCDF is loaded. Despite our .close() statement
+        # in loading code, there are still open references.
+        # os.remove(outfile)
+
```
If we're removing this, let's cut all the code not comment it out.

rstoneback replied:
I updated the comment with a TODO referencing the xarray issue. I left that code there since it is the result of time spent debugging to sort out where the error is. It is useful for the future in fully resolving #974.

aburrell commented on pysat/utils/io.py:
```diff
@@ -172,121 +167,540 @@ def filter_netcdf4_metadata(inst, mdata_dict, coltype, remove=False,
+    'Display_Type', 'Var_Type', 'Format', 'Time_Scale', 'MonoTon',
+    'calendar', and 'Time_Base'.
+
     Does not perform filtering to remove variables not supported by the
     SPDF ISTP/IACG NetCDF standards.  For this, see
     pysat.utils.io.filter_netcdf4_metadata.
```
See previous docstring for how I reworded this phrase last time.

rstoneback replied:
I updated the docstring to be more clear (hopefully). I added a See Also but referenced `add_netcdf4_standards_to_metadict` since that is the inverse of this function. The Note itself describes all of the changes to metadata.

aburrell commented on pysat/utils/io.py:
```diff
@@ -172,121 +167,540 @@ def filter_netcdf4_metadata(inst, mdata_dict, coltype, remove=False,
+                   'FillVal': meta.labels.fill_val,
+                   'fill': meta.labels.fill_val}
+
+    # We need to keep 'fill' for backwards compatibility. Unintended use
+    # of 'fill' in pysat generated files in at least v3.0.1.
```
Recommend moving this up to line 477/478, since it explains the keys above.

aburrell commented on pysat/utils/io.py:
```diff
@@ -644,8 +1136,8 @@ def load_netcdf_pandas(fnames, strict_meta=False, file_format='NETCDF4',
-                    saved_meta = meta.copy()
-                elif (meta != saved_meta):
+                    saved_meta = full_mdict.copy()
+                elif (full_mdict != saved_meta):
```
```suggestion
                elif full_mdict != saved_meta:
```

aburrell commented on pysat/utils/io.py:
```diff
@@ -724,68 +1286,256 @@ def load_netcdf_xarray(fnames, strict_meta=False, file_format='NETCDF4',
+
+    Returns
+    -------
+    variables : list
```
```suggestion
    all_vars : list
```

rstoneback replied:
snap. Thanks for catching this.

rstoneback commented on pysat/tests/test_utils_io.py:
```diff
@@ -680,75 +894,358 @@ def test_add_netcdf4_standards_to_meta(self, caplog, missing):
+
+class TestNetCDF4Integration2DXarray(TestNetCDF4Integration):
+    """Integration tests for the netCDF4 I/O utils."""
+
```
I found another class with repeated docstrings and updated that one as well.

rstoneback commented:
Thanks for the review @aburrell!

aburrell commented on pysat/tests/test_utils_io.py:
```diff
@@ -425,23 +622,34 @@ def teardown(self):
     def test_read_netcdf4_with_time_meta_labels(self, kwargs, target):
-        """Test that read_netcdf correctly interprets time labels in meta."""
-        # Prepare output test data.
+        """Test that read_netcdf correctly interprets time labels in meta.
```
```suggestion
        """Test that `read_netcdf` correctly interprets time labels in meta.
```

aburrell commented on pysat/tests/test_utils_io.py:
```diff
@@ -680,75 +894,358 @@ def test_add_netcdf4_standards_to_meta(self, caplog, missing):
+                                            ])
+    @pytest.mark.parametrize('assign_flag', [True, False])
+    def test_meta_translation_to_from_netcdf4(self, assign_flag, meta_trans):
+        """Test impact of meta_translation on netCDF output.
```
```suggestion
        """Test impact of `meta_translation` on netCDF output.
```

aburrell commented on pysat/tests/test_utils_io.py:
```diff
@@ -680,75 +894,358 @@ def test_add_netcdf4_standards_to_meta(self, caplog, missing):
+
+    @pytest.mark.parametrize('assign_flag', [True, False])
+    def test_meta_processor_to_from_netcdf4(self, assign_flag):
+        """Test impact of meta_processor on netCDF output.
```
```suggestion
        """Test impact of `meta_processor` on netCDF output.
```

aburrell commented on pysat/tests/test_utils_io.py:
```diff
@@ -680,75 +894,358 @@ def test_add_netcdf4_standards_to_meta(self, caplog, missing):
+        assign_flag : bool
+            If True, assigns meta processor func as
+            ._export_meta_post_processing. Otherwise, uses
+            meta_processor keyword argument.
```
```suggestion
            `meta_processor` keyword argument.
```

aburrell commented on pysat/tests/test_utils_io.py:
```diff
@@ -680,75 +894,358 @@ def test_add_netcdf4_standards_to_meta(self, caplog, missing):
+        ----------
+        assign_flag : bool
+            If True, assigns meta processor func as
+            ._export_meta_post_processing. Otherwise, uses
```
```suggestion
            `self._export_meta_post_processing`. Otherwise, uses
```

Is this correct?

aburrell commented on pysat/tests/test_utils_io.py:
```diff
@@ -680,75 +894,358 @@ def test_add_netcdf4_standards_to_meta(self, caplog, missing):
+        # Create meta processor function
+        to_meta_proc = functools.partial(self.meta_proc_stub, vals=target,
+                                         remove_labels=['units'])
+        # Write the file
```
```suggestion

        # Write the file
```
